File: javascript/luhn/luhn.js
export const valid = input => {
  const trimmed = input.replace(/ /g, '');
  if (trimmed.length === 1) {
    return false;
  }

  const sum = trimmed
    .split('')
    .reverse()
    .map((item, index) => {
      let newValue = Number(item);
      const evenIndex = index % 2 === 1;

      if (evenIndex) {
        newValue = newValue * 2;

        if (newValue > 9) {
          newValue = newValue - 9;
        }
      }

      return newValue;
    })
    .reduce((acc, cur) => acc + cur);

  return sum % 10 === 0;
};

File: ruby/luhn/luhn.rb
module Luhn
  extend self

  def valid?(input)
    sanitized = input.tr(' ', '')
    return false unless sanitized =~ /^[0-9]{2,}$/

    result = sanitized.chars.reverse.each_with_index.reduce(0) do |acc, (char, index)|
      integer = char.to_i

      if index.even?
        addition = integer
      else
        double = integer * 2
        addition = (double > 9 ? double - 9 : double)
      end

      acc += addition
    end

    (result % 10).zero?
  end
end
